File: linked_lists/questions/question_2.6/problem_2.6.js
var headOfCircularList = function(list){
  // debugger
  if (!list.head){
    return null;
  }

  var runners = {
    slow: list.head,
    fast: list.head
  }

  if (!collideRunners(runners)){
    return null; // non-circular list
  }
  runners.slow  = list.head;
  return haveRunnersMeet(runners)
}

var collideRunners = function(runners){
  while (runners.fast && runners.fast.next){
    runners.slow  = runners.slow.next;
    runners.fast  = runners.fast.next.next;
    if (runners.slow == runners.fast){
      return true;
    }
  }
  return false;
}

var haveRunnersMeet = function(runners){
  while(runners.slow != runners.fast){
    runners.slow  = runners.slow.next;
    runners.fast  = runners.fast.next;
  }
  return runners.slow;
}

// list = new LinkedList();
// for (var val of [1,2,3,4,5]){
//   list.push(val);
// }
// var headOfcircle = new Node(6);
// list.tail.next = headOfcircle;
// list.tail = headOfcircle;
// for (var val of [7,8,9,10,11,12,13]){
//   list.push(val);
// }
// list.tail.next = headOfcircle;

// console.log(headOfCircularList(list).value);

File: linked_lists/questions/question_2.6/problem_2.6.rb
require_relative './../../../custom_data_structures/ruby_data_structures/linked_list.rb'

class LinkedList

  def head_of_circular_list
    return nil unless @head && @head.next
    tortoise        = @head
    rabbit          = @head

    rabbit, tortoise = collide_runners(rabbit,tortoise)
    return nil unless rabbit  #acyclic list
    tortoise  = @head # now both rabbit and tortoise are k steps
                      # away from circle head
    head_of_circular_list = have_runners_meet(rabbit, tortoise)
  end

  def collide_runners(rabbit, tortoise)
    # when collision occurs, rabbit will be
    # k (length of non-cyclic front portion) steps
    # away from the circle head
    while rabbit && rabbit.next
      tortoise  = tortoise.next
      rabbit    = rabbit.next.next
      break if rabbit == tortoise
    end

    return [rabbit, tortoise] if rabbit && rabbit.next
    nil
  end

  def have_runners_meet(rabbit, tortoise)
    while true  # move both runners 1 step ahead and they will meet at the
                # circle head
      break if tortoise == rabbit
      rabbit    = rabbit.next
      tortoise  = tortoise.next
    end
    tortoise
  end
end





